### ISSUE TITLE
3.5: Improve Archived API/DB read performance

### ISSUE THREAD

**Issue by gyanprakash48** [argoproj/argo-workflows#13295]
---
_edited by agilgur5_: 
- Please help test #13566 by running [the test image below](https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2339407775)

---

### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

Currently , when large number of workflows are running , Workflows view of Argo workflows UI is taking up to 12 seconds to load. I don't find any way to further optimise it. CPU and Memory uses of server pod are normal. I am using Postgres to archive the workflows and cpu and memory usage of database as well are  normal. i don't see any spikes anywhere. So looks like its not resource contention but something else,   What can be done to improve the load time .

### Version

3.5.0, 3.5.6, 3.5.8,

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
Its UI load time (workflow view)
```

### Logs from the workflow controller

```text
Its UI load time (workflow view)
```

### Logs from in your workflow's wait container

```text
Its UI load time (workflow view)
```

**Comment by gyanprakash48** [comment]
i tried to  trace it and seem like below query is culprit . As per query performance insight(Azure Postgres flexi server) , this query execution mean  time is 9.304s, Which explains, over all api time is 12-13 seconds 

select
	name,
	namespace,
	uid,
	phase,
	startedat,
	finishedat,
	coalesce((workflow::json)->'metadata'->>'labels',
	'{}') as labels,
	coalesce((workflow::json)->'metadata'->>'annotations',
	'{}') as annotations,
	coalesce((workflow::json)->'status'->>'progress',
	'') as progress,
	coalesce((workflow::json)->'metadata'->>'creationTimestamp',
	'') as creationtimestamp,
	(workflow::json)->'spec'->>'suspend' as suspend,
	coalesce((workflow::json)->'status'->>'message',
	'') as message,
	coalesce((workflow::json)->'status'->>'estimatedDuration',
	'0') as estimatedduration,
	coalesce((workflow::json)->'status'->>'resourcesDuration',
	'{}') as resourcesduration
from
	"argo_archived_workflows"
where
	(("clustername" = $1
		and "namespace" = $2
		and "instanceid" = $3)
	and "namespace" = $4
	and not exists (
	select
		1
	from
		argo_archived_workflows_labels
	where
		clustername = argo_archived_workflows.clustername
		and uid = argo_archived_workflows.uid
		and name = 'workflows.argoproj.io/controller-instanceid'))
order by
	"startedat" desc
limit 50

**Comment by gyanprakash48** [comment]
These json functions are culprit, coalesce((workflow::json)->'metadata'->>'labels',
'{}') as labels,
coalesce((workflow::json)->'metadata'->>'annotations',
'{}') as annotations,
coalesce((workflow::json)->'status'->>'progress',
'') as progress,
coalesce((workflow::json)->'metadata'->>'creationTimestamp',
'') as creationtimestamp,
(workflow::json)->'spec'->>'suspend' as suspend,
coalesce((workflow::json)->'status'->>'message',
'') as message,
coalesce((workflow::json)->'status'->>'estimatedDuration',
'0') as estimatedduration,
coalesce((workflow::json)->'status'->>'resourcesDuration',

**Comment by driv** [comment]
I've been testing the queries from the logs of `3.5.8` and I can see a huge difference between MySQL 5.7 and 8.

The queries execution plans are difference and on 8 it executes more than 2 orders of magnitude faster.

I'm attaching the queries EXPLAIN.

[explain.57.md](https://github.com/user-attachments/files/16087324/explain.57.md)
[explain.8.md](https://github.com/user-attachments/files/16087325/explain.8.md)

MySQL 8 is able to use range.

**Comment by gyanprakash48** [comment]
hmm, i am using Postgres Flexi server , so not sure if it comparable.  
What about storing these  columns as generated column in the table,  and in api  query using those generated columns.  something like below. I tested it for same data where currently it is taking 9 sec, revised query finishing in milliseconds. Will it have any downside ?  Get Workflows  route is most frequently used  in UI  and its slowness taking away all the charm argo workflows (Users always feel frustrated).
 table DDL
=======
CREATE TABLE public.argo_archived_workflows (
	uid varchar(128) NOT NULL,
	"name" varchar(256) NOT NULL,
	phase varchar(25) NOT NULL,
	"namespace" varchar(256) NOT NULL,
	workflow json NOT NULL,
	startedat timestamp DEFAULT CURRENT_TIMESTAMP NOT NULL,
	finishedat timestamp DEFAULT CURRENT_TIMESTAMP NOT NULL,
	clustername varchar(64) NOT NULL,
	instanceid varchar(64) NOT NULL,
	labels text GENERATED ALWAYS AS ((workflow::json)->'metadata'->>'labels') stored,
	annotations text GENERATED ALWAYS as ( (workflow::json)->'metadata'->>'annotations') stored,
	progress text GENERATED ALWAYS AS ((workflow::json)->'status'->>'progress') stored,
	creationtimestamp text GENERATED ALWAYS AS ((workflow::json)->'metadata'->>'creationTimestamp') stored,
	suspend text GENERATED ALWAYS AS ((workflow::json)->'spec'->>'suspend') stored,
	message text GENERATED ALWAYS AS ((workflow::json)->'status'->>'message') stored,
	estimatedduration text GENERATED ALWAYS AS ((workflow::json)->'status'->>'estimatedDuration') stored,
	resourcesduration text GENERATED ALWAYS AS ((workflow::json)->'status'->>'resourcesDuration') stored,  
	CONSTRAINT argo_archived_workflows_pkey PRIMARY KEY (clustername, uid)
);
CREATE INDEX argo_archived_workflows_i1 ON public.argo_archived_workflows USING btree (clustername, instanceid, namespace);
CREATE INDEX argo_archived_workflows_i2 ON public.argo_archived_workflows USING btree (clustername, instanceid, finishedat);
CREATE INDEX argo_archived_workflows_i3 ON public.argo_archived_workflows USING btree (clustername, instanceid, name);
CREATE INDEX argo_archived_workflows_i4 ON public.argo_archived_workflows USING btree (startedat);

Revised Query 
=======

select
	name,
	namespace,
	uid,
	phase,
	startedat,
	finishedat,
	coalesce(labels,'{}') as labels,
	coalesce(annotations,'{}') as annotations,
	coalesce(progress,'') as progress,
	coalesce(creationtimestamp,'') as creationtimestamp,
	suspend,
	coalesce(message,'') as message,
	coalesce(estimatedduration,'0') as estimatedduration,
	coalesce(resourcesduration,'{}') as resourcesduration
from
	public.argo_archived_workflows_test 
where
	(("clustername" = $1
		and "namespace" = $2
		and "instanceid" = $3)
	and "namespace" = $4
	and not exists (
	select
		1
	from
		argo_archived_workflows_labels
	where
		clustername = 'default'
		and uid = argo_archived_workflows_test.uid
		and name = 'workflows.argoproj.io/controller-instanceid'))
order by startedat desc
limit 45;

**Comment by driv** [comment]
Given that the query has a limit 50 and those fields are not used in filtering or sorting, I doubt they are responsible for the long query duration.

Can you share an EXPLAIN?

**Comment by gyanprakash48** [comment]
in the beginning i though that too, but  trying different strategy to switch  different explain plan was not helping at all . Then i comment out those json  function column in select and suddenly query was super fast.

**Comment by gyanprakash48** [comment]
![log](https://github.com/argoproj/argo-workflows/assets/109652198/305896fa-6a20-42d9-b0b7-517782bd838e)
May be this recording will  help further. I am sure this is only one of the  of the scenario (must be many others)

**Comment by agilgur5** [comment]
> Then i comment out those json function column in select, and suddenly query was super fast. Even Though they are not part of sorting and filtering, but these json are very big in size(If workflows is large , at least thats the case in my situation)

Yea they can be very large, though I'm surprised it takes that long, since it's not part of the filter. I'm wondering if the engine is extracting the JSON multiple times, once for each field?

> I tested it for same data where currently it is taking 9 sec, revised query completing in milliseconds. does it have any downside ?

> ```sql
> 	labels text GENERATED ALWAYS AS ((workflow::json)->'metadata'->>'labels') stored,
> 	annotations text GENERATED ALWAYS as ( (workflow::json)->'metadata'->>'annotations') stored,
> 	progress text GENERATED ALWAYS AS ((workflow::json)->'status'->>'progress') stored,
> 	creationtimestamp text GENERATED ALWAYS AS ((workflow::json)->'metadata'->>'creationTimestamp') stored,
> 	suspend text GENERATED ALWAYS AS ((workflow::json)->'spec'->>'suspend') stored,
> 	message text GENERATED ALWAYS AS ((workflow::json)->'status'->>'message') stored,
> 	estimatedduration text GENERATED ALWAYS AS ((workflow::json)->'status'->>'estimatedDuration') stored,
> 	resourcesduration text GENERATED ALWAYS AS ((workflow::json)->'status'->>'resourcesDuration') stored,
> ```

We don't currently filter on those, so it's a little unnecessary as individual columns, but otherwise that makes sense to me if it's compatible with MySQL as well. I'm wondering if we can just use a materialized view instead for similar effect.

cc @jiachengxu @terrytangyuan @jessesuen

**Comment by jlpgz** [comment]
We are experiencing the same problem using PostgreSQL. There are nearly 1 million archived workflows in the database, making the query extremely expensive to run. As a result, the UI becomes unusable due to its slow response time.

I have always wondered why the controller does not create a JSONB column to store the workflow information.

**Comment by jlpgz** [comment]
Would it be possible to change the `workflow` column to JSONB in Postgres via migration? JSON operations would be faster and we could use GIN indexes.

**Comment by jiachengxu** [comment]
> Would it be possible to change the `workflow` column to JSONB in Postgres via migration? JSON operations would be faster and we could use GIN indexes.

I think this is a good point, currently for Postgres, we use [json](https://github.com/argoproj/argo-workflows/blob/main/persist/sqldb/migrate.go#L199-L202) but not JSONB. In the archived workflow case(heavy read), I think it makes sense to use JSONB since:
> The json data type stores an exact copy of the input text, which processing functions must reparse on each execution; while jsonb data is stored in a decomposed binary format that makes it slightly slower to input due to added conversion overhead, but significantly faster to process, since no reparsing is needed. jsonb also supports indexing, which can be a significant advantage.

https://www.postgresql.org/docs/current/datatype-json.html

> Yea they can be very large, though I'm surprised it takes that long, since it's not part of the filter. I'm wondering if the engine is extracting the JSON multiple times, once for each field?

@agilgur5 To your point, from the above quote "while jsonb data is stored in a decomposed binary format that makes it slightly slower to input due to added conversion overhead, but significantly faster to process, since no reparsing is needed", I think for current implementation, most likely it is reparsing the json for each field.

**Comment by agilgur5** [comment]
> and we could use GIN indexes.

We don't need indexes on it currently as the JSON is not used in filtering.

> I have always wondered why the controller does not create a JSONB column to store the workflow information.

Otherwise, it does sound like JSONB is the better choice for our read queries. This is only supported for Postgres though, and we'd have to add the min version of Postgres required for JSONB to the docs. I imagine lack of MySQL support and legacy code is the reason but 🤷

Is it straightforward to write a migration to JSONB?

We'd also still have to solve this for MySQL, so that alone wouldn't be enough. A materialized view or generated columns are probably needed, and compressed nodes (#13313) might help too.

> I think for current implementation, most likely it is reparsing the json for each field.

That would explain why it may have gotten slower. That is odd to me that it's reparsed (as the engine should be able to store it in memory and then reuse it for the query), but materialized views / generated columns should solve that relatively cleanly, assuming that writes are not heavily slowed down (e.g. if it reparses each field during derivation/generation too)

**Comment by hmoulart** [comment]
Hello,
I have the same behavior on that part with the database overloaded with archiveTTL: 180d
Add an index on some columns was not enough...
<img width="1276" alt="Capture d’écran 2024-07-03 à 12 20 23" src="https://github.com/argoproj/argo-workflows/assets/16108020/15fa3a89-a8a1-400f-a2cd-3bdb352f686c">
database is a AWS RDS PostgreSQL t4g.small
<img width="451" alt="image" src="https://github.com/argoproj/argo-workflows/assets/16108020/adac50a8-9c18-48d8-af66-c1a4bbe73665">
archived_workflows: 816 757
argo_archived_workflows: 37 427
Number of line is not so huge...

We created a job on PostgreSQL database to keep 2 days on the active argo table and 178d on the archive argo table. The situation is now better.
But Argo hit the database very often for no reason with the query you can see on the screen (which is the same as the query on this topic).
I'm available to help identify the issue and optimize Argo to avoid overloading databases.
Thanks for your support!

**Comment by jlpgz** [comment]
> Is it straightforward to write a migration to JSONB?
I would say yes. My only concern would be how long can take to convert the workflow column in a large number of rows.

```sql
ALTER TABLE argo_archived_workflows
ALTER COLUMN workflow
SET DATA TYPE JSONB
USING workflow::JSONB;
```

**Comment by gyanprakash48** [comment]
changing to type JSONB solves the performance issue completely ?   someone has validated it ?

**Comment by Oscmage** [comment]
Hi, we are getting a similar issue with a slow query + high CPU usage using postgres 15.7 for the query:

```
SELECT
  name,
  namespace,
  uid,
  phase,
  startedat,
  finishedat,
  COALESCE((workflow::json)->$4->>$5, $6) AS labels,
  COALESCE((workflow::json)->$7->>$8, $9) AS annotations,
  COALESCE((workflow::json)->$10->>$11, $12) AS progress,
  COALESCE((workflow::json)->$13->>$14, $15) AS creationtimestamp,
  (workflow::json)->$16->>$17 AS suspend,
  COALESCE((workflow::json)->$18->>$19, $20) AS message,
  COALESCE((workflow::json)->$21->>$22, $23) AS estimatedduration,
  COALESCE((workflow::json)->$24->>$25, $26) AS resourcesduration
FROM
  "argo_archived_workflows"
WHERE
  (("clustername" = $1
      AND "instanceid" = $2)
    AND "namespace" = $3
    AND EXISTS (
    SELECT
      $27
    FROM
      argo_archived_workflows_labels
    WHERE
      clustername = argo_archived_workflows.clustername
      AND uid = argo_archived_workflows.uid
      AND name = $28
      AND value = $29)
    AND EXISTS (
    SELECT
      $30
    FROM
      argo_archived_workflows_labels
    WHERE
      clustername = argo_archived_workflows.clustername
      AND uid = argo_archived_workflows.uid
      AND name = $31
      AND value = $32))
ORDER BY
  "startedat" DESC
LIMIT
  $33
```

![Screenshot 2024-07-26 at 10 53 48](https://github.com/user-attachments/assets/d0a83f9a-2e18-444b-863f-dfdd10800f3c)

**Comment by agilgur5** [comment]
> > Is it straightforward to write a migration to JSONB?
> 
> I would say yes. My only concern is how long it might take to convert the workflow column for a large number of rows.

Yes good point. If this does take long, we should probably defer that to 3.6 so that it can be noted in the upgrading guide. Would someone like to test it?

Similarly, would someone like to test/PoC for a materialized view?

**Comment by rhoens** [comment]
Also, note that in the query there are a bunch of casts to `::json`. Removing these is a relatively significant win on queries with a `limit` clause (e.g., on `limit 200` my queries went from 5seconds to 2 seconds).

**Comment by gyanprakash48** [comment]
Would not GENERATED column approach be easy win,  as other perfect  solution  yet to be discovered. ?

**Comment by agilgur5** [comment]
> Also, note that in the query there are a bunch of casts to `::json`. Removing these is a relatively significant win

Yes, per the comments above, that's what the JSONB migration is intended to improve performance of

> queries with a `limit` clause (e.g., on `limit 200` my queries went from 5seconds to 2 seconds).

Afaik, the only time a query has no `limit` is quite literally when the UI or another API client requests a no limit list. So we can't really prevent that per se and that isn't unique to 3.5 either.

> and was able to get it to complete,

> It still takes minutes/hours to run this query (I'm still waiting for it to finish after ~10 mins)

> It did take quite a few minutes to actually do the conversion

Sorry, I'm a little confused here -- did the migration finish or not? Not sure if you were doing it in multiple environments, but some of your sentences say it completed and another says its still running. If it's still running, that would certainly affect any queries.

> returns: `73,046`

Thanks for testing and providing the timing and size details. That's a comparatively small amount as I know users with easily millions of rows. So that would suggest that this migration would indeed be a big hit that may make sense to hold off on until a minor release for a proper upgrade note 😕 

I'm thinking a materialized view might have a similar issue, but we could potentially write fallback code to workaround it.
Annnd while writing this I learned that MySQL also [doesn't support materialized views](https://dba.stackexchange.com/a/86793/299298). And Postgres apparently requires them to be manually refreshed as well 😕  AWS has a [nice table](https://aws.amazon.com/what-is/materialized-view/) on this (under "Technical variations in different systems").

It seems like the only approach compatible with both Postgres and MySQL is the [generated columns approach above](https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2206733669). Stored columns might cause a longer migration time as well though. And I'm not sure if the [minimum versions Argo specifies](https://argo-workflows.readthedocs.io/en/release-3.5/workflow-archive/) support generated columns (I would hope so, but 🤷 at this point)

**Comment by rhoens** [comment]
> Sorry, I'm a little confused here -- did the migration finish or not? Not sure if you were doing it in multiple environments, but some of your sentences say it completed and another says it's still running. If it's still running, that would certainly affect any queries.
> 
> >

Sorry for the confusion. The migration completed. So now the table is using `jsonb` for workflows.

After the migration, I reran the query (without the limit), and there was no change in performance. It still took hours to complete a single query, and even started taking down the DB. I had to revert to `3.4.8` which reduced load on the DB significantly.

>I'm thinking a materialized view might have a similar issue, but we could potentially write fallback code to workaround it.
My recommendation here is to just widen the table. It seems that, for whatever reason, postgres really doesn't like giant json blobs. If you splat out the queried fields, my assumption is it would be much faster. I haven't been able to test this, as I was dealing with the DB going down, but that's my recommendation based on this.

**Comment by jlpgz** [comment]
But the query is still converting the workflow column to json, isn’t it?

**Comment by agilgur5** [comment]
> Sorry for the confusion. The migration completed. So now the table is using `jsonb` for workflows.
> 
> After the migration, I reran the query (without the limit), and there was no change in performance

Ok thanks for clarifying. I imagine that without a limit any query is going to be slow. JSONB should still be faster than the JSON variant regardless though.

> But the query is still converting the workflow column to json, isn’t it?

Is there a separate query for JSONB specifically? I actually don't know

> My recommendation here is to just widen the table. It seems that, for whatever reason, postgres really doesn't like giant json blobs. If you splat out the queried fields, my assumption is it would be much faster.

Yea the [generated columns approach](https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2206733669) is more or less equivalent to that and was significantly faster (order of magnitude or two) per testing in that comment.

**Comment by agilgur5** [comment]
So I've been thinking about this after discussing in the Aug 6th Contributor Meeting with @sarabala1979. All of the options we discussed to support this had trade-offs, so no optimal solution.

I was thinking about the behavior of the Archived Workflows API prior to #11121, which takes me back to my own comment, https://github.com/argoproj/argo-workflows/pull/11121#discussion_r1469101802. The fix to that comment was https://github.com/argoproj/argo-workflows/pull/12912 which added the `JSON_EXTRACT` statements instead of retrieving the entirety of the JSON, which caused _both_ slow queries and OOMs on the Server. Now it's just a slow query at least 😅

We could remove those fields entirely from the `SELECT`, i.e. never return from the Archived Workflows List API, and therefore never need to extract the JSON or add new columns. This would reflect the behavior of 3.4, but would make for some confusing inconsistencies in the unified list in 3.5+, as some items of the list would have the fields, but the archived ones would not.
This seems to not strongly affect users though afaict, because https://github.com/argoproj/argo-workflows/issues/13098 did not get any upvotes and is similar.
The main problem with this approach is that it limits the types of filtering that can be done on archived workflows (e.g. labels are extracted from the JSON), and we expected to be able to implement more (more have always been requested) after the SQLiteDB addition in #12736 / #13021. Creating new columns for every piece of metadata users want to filter on is not necessarily sustainable either though.
So, I think if necessary, we can rollback to that behavior, and then see what we can do in the future in 3.6+.

But perhaps #13566 has the root cause fix if all the errors are worked out and we won't have to think about any more options

**Comment by ryancurrah** [comment]
> We could remove those fields entirely from the SELECT, i.e. never return from the Archived Workflows List API, and therefore never need to extract the JSON or add new columns. This would reflect the behavior of 3.4, but would make for some confusing inconsistencies in the unified list in 3.5+, as some items of the list would have the fields, but the archived ones would not.

Not sure if I fully understand this comment. We search archived workflows by labels frequently because we archive workflows almost immediately. Would that mean could not do that anymore?

**Comment by agilgur5** [comment]
I did forget a piece in my initial comment, which was that labels actually have their own table already, so it would only be other forms of information in the JSON blob that would not be filterable, like annotations, pieces of the `status`, etc. Whatever doesn't have an existing column

**Comment by agilgur5** [comment]
> But perhaps #13566 has the root cause fix if all the errors are worked out and we won't have to think about any more options

For other folks here who haven't been following the PR, I pushed up [`agilgur5/argo-cli:fix-archive-list-query-optimization`](https://hub.docker.com/layers/agilgur5/argo-cli/fix-archive-list-query-optimization/images/sha256-d3a31a31e051a499e7b49e93a7c942414cf83eb56d6e46cc012b23758e1b36a7?context=repo) for testing.

Please try it out and comment here with your query improvement results! (or lack thereof if it doesn't work for you)

**Comment by agilgur5** [comment]
From @Danny5487401 in https://github.com/argoproj/argo-workflows/issues/13563#issuecomment-2339660938 

> @agilgur5 three methods i tested:
> 
> 1. nice implemention by sub,according to [fix(api): optimise archived list query. Fixes #13295 #13566](https://github.com/argoproj/argo-workflows/pull/13566)
> 
> ```sql
> SELECT
>     name, namespace, uid, phase, startedat, finishedat,
>     coalesce(JSON_EXTRACT(workflow,'$.metadata.labels'), '{}') as labels,
>     coalesce(JSON_EXTRACT(workflow,'$.metadata.annotations'), '{}') as annotations,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.status.progress')), '') as progress,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.metadata.creationTimestamp')), '') as creationtimestamp,
>     JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.spec.suspend')) as suspend,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.status.message')), '') as message,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.status.estimatedDuration')), '0') as estimatedduration,
>     coalesce(JSON_EXTRACT(workflow,'$.status.resourcesDuration'), '{}') as resourcesduration
> FROM `argo_archived_workflows`
> WHERE `clustername` = 'default' AND uid IN
> ( SELECT * FROM (SELECT uid
> FROM `argo_archived_workflows`
> WHERE ((`clustername` = 'default' AND `instanceid` = '') AND `namespace` = 'gllue-web'
>     AND not exists
>     ( select 1 from argo_archived_workflows_labels where clustername = argo_archived_workflows.clustername and uid = argo_archived_workflows.uid and name = 'workflows.argoproj.io/controller-instanceid')
>     AND exists
>     ( select 1 from argo_archived_workflows_labels where clustername = argo_archived_workflows.clustername and uid = argo_archived_workflows.uid and name = 'workflows.argoproj.io/phase' and value in ('Error', 'Failed', 'Pending', 'Running', 'Succeeded'))
>           )
> ORDER BY `startedat` DESC
>     LIMIT 20)as x);
> ```
> 
> it drops from 37s 768ms to 1 s 981 ms
> 
> 2. force index (argo_archived_workflows_i4)
> 
> ```sql
> SELECT
>     name, namespace, uid, phase, startedat, finishedat,
>     coalesce(JSON_EXTRACT(workflow,'$.metadata.labels'), '{}') as labels,
>     coalesce(JSON_EXTRACT(workflow,'$.metadata.annotations'), '{}') as annotations,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.status.progress')), '') as progress,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.metadata.creationTimestamp')), '') as creationtimestamp,
>     JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.spec.suspend')) as suspend,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.status.message')), '') as message,
>     coalesce(JSON_UNQUOTE(JSON_EXTRACT(workflow,'$.status.estimatedDuration')), '0') as estimatedduration,
>     coalesce(JSON_EXTRACT(workflow,'$.status.resourcesDuration'), '{}') as resourcesduration
> FROM `argo_archived_workflows` force index (argo_archived_workflows_i4)
> WHERE ((`clustername` = 'default' AND `instanceid` = '') AND `namespace` = 'gllue-web' AND not exists
>     ( select 1 from argo_archived_workflows_labels where clustername = argo_archived_workflows.clustername and uid = argo_archived_workflows.uid and name = 'workflows.argoproj.io/controller-instanceid')
>            AND exists
>     ( select 1 from argo_archived_workflows_labels where clustername = argo_archived_workflows.clustername and uid = argo_archived_workflows.uid and name = 'workflows.argoproj.io/phase' and value in ('Error', 'Failed', 'Pending', 'Running', 'Succeeded'))
>     )
> ORDER BY `startedat` DESC
> LIMIT 20;
> ```
> 
> it drops from 37s 768ms to 138 ms
> 
> 3. union index (clustername, startedat) to use Backward index scan instead of filesort
> 
> ```sql
> CREATE INDEX argo_archived_workflows_i5 ON argo_archived_workflows (clustername, startedat);
> ```
> 
> it drops from 37s to 140 ms

It sounds like the `force index` and new index are both faster than the subquery, at least on MySQL 8.0, I wonder if we should use those instead of the subquery

**Comment by ryancurrah** [comment]
@agilgur5 the image name is `argo-cli` can I use it to replace argo-server and argo-executor?

**Comment by ryancurrah** [comment]
I can confirm this change reduces query times. Loading time in the UI on filtering labels was reduced by 94.44%. 

**Query time Before**: 18 seconds
**Query time after** : 1 second
**DB type**: Postgresql
**Version**: 14
**Number of rows (`SELECT COUNT(*) FROM public.argo_archived_workflows;`)**: 93135
**CPUs**: 24
**Memory**: 16GB

**Comment by agilgur5** [comment]
> It sounds like the `force index` and new index are both faster than the subquery, at least on MySQL 8.0, I wonder if we should use those instead of the subquery.
> `force index` is potentially not the most resilient, but the new index seems like a "why not?" to me

Added a new issue for further optimizations to this query in #13601

**Comment by agilgur5** [comment]
> sorry for delay from my side in testing, but even though with `agilgur5/argo-cli:fix-archive-list-query-optimization`

It's already released in 3.5.11. 

> i am not finding any improvement in UI performance

But you'd really have to detail why you're not seeing any improvement in API performance when other people are. Did you run the image for your Server properly? What kinds of queries are you using to the API and seeing in the DB? etc

Per https://github.com/argoproj/argo-workflows/issues/13601#issuecomment-2377275101, unpaginated queries can still be slow (indexing doesn't help that of course)

**Comment by gyanprakash48** [comment]
@agilgur5  , yes, i applied the image  (agilgur5/argo-cli:fix-archive-list-query-optimization) for server  and then for testing ,i simply use the UI. UI was still loading very slow (its paginated and limit  50 , default). Only difference for my use case might be , my workflows are of very large size (1000 + task) leading to very large size json file
